[ThinLTO] Don't convert functions to declarations if `force-import-all` is enabled

shiltian opened the pull request:
On one hand, we intend to force import all functions when the option is enabled.
On the other hand, we currently drop definitions of some functions and convert
them to declarations, which contradicts this intent.

With this PR, functions will no longer be converted to declarations when
`force-import-all` is enabled.

shiltian commented:
* **#134541** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134541?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134541?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

cachemeifyoucan commented on llvm/test/ThinLTO/AMDGPU/force-import-all.ll:
```diff
@@ -0,0 +1,35 @@
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -module-summary %s -o %t.main.bc
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -module-summary %p/Inputs/in-0.ll -o %t.in.0.bc
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -module-summary %p/Inputs/in-1.ll -o %t.in.1.bc
```
I am not quite sure what value does the second module add for the test?

cachemeifyoucan commented on llvm/test/ThinLTO/AMDGPU/force-import-all.ll:
```diff
@@ -0,0 +1,35 @@
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -module-summary %s -o %t.main.bc
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -module-summary %p/Inputs/in-0.ll -o %t.in.0.bc
+; RUN: opt -mtriple=amdgcn-amd-amdhsa -module-summary %p/Inputs/in-1.ll -o %t.in.1.bc
+; RUN: llvm-lto -thinlto-action=run -force-import-all %t.main.bc %t.in.0.bc %t.in.1.bc --thinlto-save-temps=%t.2.
```
Don't you just need to run `-thinlto-action=import` instead of running the full pipeline and rely on save-temps?

shiltian replied:
I wanted to do that as well, but it looks like `thinLTOResolvePrevailingGUID` doesn't run if we simply does `import`, as shown below:

```
; RUN: llvm-lto -thinlto-action=thinlink -force-import-all %t.main.bc %t.in.bc -o %t.index.bc
; RUN: llvm-lto -thinlto-action=import -force-import-all %t.main.bc %t.in.bc --thinlto-index=%t.index.bc
; RUN: llvm-dis %t.main.bc.thinlto.imported.bc -o - | FileCheck --check-prefix=MOD1 %s
; RUN: llvm-dis %t.in.bc.thinlto.imported.bc -o - | FileCheck --check-prefix=MOD2 %s
```

Did I miss anything here?

cachemeifyoucan commented on llvm/lib/Transforms/IPO/FunctionImport.cpp:
```diff
@@ -1701,13 +1701,16 @@ void llvm::thinLTOFinalizeInModule(Module &TheModule,
     if (GlobalValue::isAvailableExternallyLinkage(NewLinkage) &&
-        GlobalValue::isInterposableLinkage(GV.getLinkage())) {
+        GlobalValue::isInterposableLinkage(GV.getLinkage()) &&
+        !ForceImportFunction) {
```
Isn't the comment above explained why such function should not be inlined? Converting to available externally linkage is changing the behavior of the interposable linkage, and only a declaration (undef) can preserve the behavior.

I don't think a `-mllvm` option (even it is debugging option) should violate the linkage behavior for the IR language.

shiltian replied:
First of all, `force-import-all` is not a debug option. It was introduced to support targets that don’t support object linking, like AMDGPU.

Second, while the comment above explains the reasoning for this code block, it assumes that object linking is supported. For targets that don’t support object linking, none of what’s described in the comment will actually happen.

shiltian replied:
> All `cl::opt` in the passes are all considered not officially supported user facing options. You need change that to an updated the actual `lto::Config` to has an option for that to be an official function.

Sure, this will be done in a separate PR.

> The comment explains that for an `interposable linkage`, it is invalid to be inlined according to language ref. It is not about if object linking is allowed or not. You might use a wrong linkage type here? Maybe something `linkonce_odr` is better if inlining should be allowed (of course I don't know your use case)

The global is set to `available_externally` if it is not prevailing nor an alias in `thinLTOResolvePrevailingGUID`. This might need to be updated to consider the force-import-all case.

cachemeifyoucan replied:
Using your test case as example, you have:
main.ll: f0, weak_common (both prevailing)
1.ll: f1 (prevailing), weak_common (not prevailing)

After importing:
main.ll: f0, f1(available_externally), weak_common (prevailing)
1.ll: f1 (prevailing), weak_common (available_externally)

The possible optimization:
main.ll: f0, weak_common (prevailing), f1 is not inlined and discarded
1.ll: f1 (prevailing), weak_common is inlined.

Result:
f0 calls f1, which has an inlined version of weak_common, which lost interposablility.

cachemeifyoucan replied:
The goes back to my question before, that is if AMD GPU should use other linkage than `weak`? If you want to allow inlining in all cases and interposability is not going to happen, maybe use `weak_odr`? The better fix is probably fix the IRGen to use `weak_odr` in the places of `weak`, rather than introduce a spec violation with this flag.

shiltian replied:
FWIW, the goal here isn't about enabling inlining. What we actually want is to ensure that function definitions exist in every module, regardless of whether the functions get inlined or not. They just can't be converted to declarations.

In fact, the specific case I'm trying to resolve in this PR is a function defined in a header file like this:

```cpp
extern "C" __device__ __attribute__((noinline)) __attribute__((weak)) void __assert_fail(...
```

This function is marked `noinline`, so it won't be inlined. However, it still ends up being converted into a declaration, which is the issue we're trying to fix.

cachemeifyoucan replied:
I am not familiar with AMDGPU backend setup and how device function is modeled in LLVM IR. How does available_externally linkage behave in AMDGPU Codegen? Does it still get dropped if not used, and do you need to prevent it from dropped (or you just need them in IR for some reason)?

If the goal is to have all definitions, maybe we should avoid function importer getting to `available_externally` linkage from the first place and keep weak function as weak. I am not sure if you need to also prevent them from turning into available externally in other linkage type.

shiltian replied:
> I am not familiar with AMDGPU backend setup and how device function is modeled in LLVM IR. How does available_externally linkage behave in AMDGPU Codegen? Does it still get dropped if not used, and do you need to prevent it from dropped (or you just need them in IR for some reason)?

Yes, that's a good point. It is indeed dropped and that's why I have another [PR](https://github.com/llvm/llvm-project/pull/134476) to prevent it from getting dropped by the pass.

> If the goal is to have all definitions, maybe we should avoid function importer getting to available_externally linkage from the first place and keep weak function as weak. I am not sure if you need to also prevent them from turning into available externally in other linkage type.

For this particular case regarding `__assert_fail`, it is not imported actually since each TU has its own definition. However, it is set to `available_externally` for all non-prevailing ones in `thinLTOResolvePrevailingGUID` as I mentioned earlier.

For the other function that are indeed imported, yes, we might want to consider to change to something else.

cachemeifyoucan replied:
Ah, that has the context I needed to understand your PR. And you rely on the `avail-extern-to-local` later to make all function definition as local, so they are all available.

From a high level design point of view, I would prefer collapse `avail-extern-to-local` and `force-import-all` to one option, and instead of running two passes (that both alter the semantics of the IR), just teach function importer to do the right thing as one step.

shiltian replied:
I mean, that is for an imported symbol, but the `weak_common` in the newly added test case is not an imported symbol, since every module has its own definition?

cachemeifyoucan replied:
My suggestion is to prevent `weak_common` turning into available_externally in the first place (I think it happens in `thinLTOResolvePrevailingGUID`).

shiltian committed:
[ThinLTO] Don't convert functions to declarations if `force-import-all` is enabled

On one hand, we intend to `force-import-all` functions when the option is
enabled. On the other hand, we currently drop definitions of some functions and
convert them to declarations, which contradicts this intent.

With this PR, functions will no longer be converted to declarations when
`force-import-all` is enabled.

shiltian commented:
@cachemeifyoucan How about this version?

cachemeifyoucan reviewed (approved):
LGTM. Thanks!
